core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java

```diff
@@ -33,16 +31,22 @@ public class ClusterTestExtensionsUnitTest {
     void testProcessClusterTemplate() {
         ClusterTestExtensions ext = new ClusterTestExtensions();
         ExtensionContext context = mock(ExtensionContext.class);
-        Consumer<TestTemplateInvocationContext> testInvocations = mock(Consumer.class);
         ClusterTemplate annot = mock(ClusterTemplate.class);
-        when(annot.value()).thenReturn("").thenReturn(" ");
+        when(annot.value()).thenReturn("").thenReturn(" ").thenReturn("test_empty_config");
```

Comment:
I'm wondering if these tests would be simpler to write with actual methods. 

We could define a static inner class:

```java
static class StubTest {
    @ClusterTemplate("cfgFoo")
    void testFoo() {}
    static void cfgFoo(ClusterGenerator gen) { /* ... */ }

    @ClusterTemplate("")
    void testBar() {}
}
```

and a utility method:

```java
private ExtensionContext buildExtensionContext(String methodName) throws Exception {
    ExtensionContext extensionContext = mock(ExtensionContext.class);
    Class clazz = StubTest.class;
    Method method = clazz.getDeclaredMethod(methodName);
    when(extensionContext.getRequiredTestClass()).thenReturn(clazz);
    when(extensionContext.getRequiredTestMethod()).thenReturn(method);
    return extensionContext;
}
```

and then we could test it this way:

```java
ClusterTestExtensions clusterTestExtensions = new ClusterTestExtensions();

assertEquals("ClusterConfig generator method should provide at least one config.",
        assertThrows(IllegalStateException.class, () ->
                clusterTestExtensions.provideTestTemplateInvocationContexts(buildExtensionContext("testFoo"))
        ).getMessage()
);

assertEquals("ClusterTemplate value can't be empty string.",
        assertThrows(IllegalStateException.class, () ->
                clusterTestExtensions.provideTestTemplateInvocationContexts(buildExtensionContext("testBar"))
        ).getMessage()
);

/* etc */
```

This would also mean that we can keep `processClusterTemplate` private and test `provideTestTemplateInvocationContexts` directly.

Reply:
That is a good way. Also, we can do the refactor in https://issues.apache.org/jira/browse/KAFKA-16654 later.

Reply:
@soarez Thanks for your comments. All is vert excellent!!! Thanks!

Reply:
@soarez It seems not work because tester will test `@ClusterTemplate("cfgFoo")` and `@ClusterTemplate("")` and these method will make test fail.
Please look at https://ge.apache.org/s/yyivyrph2oas6, thanks.

Reply:
ok, we have many checks to make the invalid annotation failed. Maybe we go back to test each helper methods instead?